hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -155,46 +155,68 @@ int SaraNcpClient::initParser(Stream* stream) {
-        int r = CHECK_PARSER_URC(reader->scanf("+CREG: %d,%d", &val[0], &val[1]));
+        int val[4] = {-1,-1,-1,-1};
+        // Parse as direct response (ignoring mode)
+        int r = CHECK_PARSER_URC(reader->scanf("+CREG: %*d,%d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3]));
+        // Reparse as URC
+        if (r <= 1) { r = CHECK_PARSER_URC(reader->scanf("+CREG: %d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3])); }
         CHECK_TRUE(r >= 1, SYSTEM_ERROR_UNKNOWN);
         // Home network or roaming
-        if (val[r - 1] == 1 || val[r - 1] == 5) {
+        if (val[0] == 1 || val[0] == 5) {
             self->creg_ = RegistrationState::Registered;
         } else {
             self->creg_ = RegistrationState::NotRegistered;
         }
         self->checkRegistrationState();
+        // Cellular Global Identity (partial)
+        self->cgi_.location_area_code = static_cast<uint16_t>(val[2]);
+        self->cgi_.cell_id = val[3];
         return 0;
     }, this));
+    // n={0,1} +CGREG: <stat>
+    // n=2     +CGREG: <stat>[,<lac>,<ci>[,<AcT>,<rac>]]
     CHECK(parser_.addUrcHandler("+CGREG", [](AtResponseReader* reader, const char* prefix, void* data) -> int {
         const auto self = (SaraNcpClient*)data;
-        int val[2];
-        int r = CHECK_PARSER_URC(reader->scanf("+CGREG: %d,%d", &val[0], &val[1]));
+        int val[4];
+        // Parse as direct response (ignoring mode)
+        int r = CHECK_PARSER_URC(reader->scanf("+CGREG: %*d,%d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3]));
+        // Reparse as URC
+        if (r <= 1) { r = CHECK_PARSER_URC(reader->scanf("+CGREG: %d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3])); }
         CHECK_TRUE(r >= 1, SYSTEM_ERROR_UNKNOWN);
         // Home network or roaming
-        if (val[r - 1] == 1 || val[r - 1] == 5) {
+        if (val[0] == 1 || val[0] == 5) {
             self->cgreg_ = RegistrationState::Registered;
         } else {
             self->cgreg_ = RegistrationState::NotRegistered;
         }
         self->checkRegistrationState();
+        // Cellular Global Identity (partial)
+        self->cgi_.location_area_code = val[2];
+        self->cgi_.cell_id = val[3];
         return 0;
     }, this));
+    // +CEREG: <stat>[,[<tac>],[<ci>],[<AcT>][,<cause_type>,<reject_cause>[,[<Active_Time>],[<Periodic_TAU>]]]]
     CHECK(parser_.addUrcHandler("+CEREG", [](AtResponseReader* reader, const char* prefix, void* data) -> int {
         const auto self = (SaraNcpClient*)data;
-        int val[2];
-        int r = CHECK_PARSER_URC(reader->scanf("+CEREG: %d,%d", &val[0], &val[1]));
+        int val[4];
+        // Parse as direct response (ignoring mode)
+        int r = CHECK_PARSER_URC(reader->scanf("+CEREG: %*d,%d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3]));
+        // Reparse as URC
+        if (r <= 1) { r = CHECK_PARSER_URC(reader->scanf("+CEREG: %d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3])); }
```

Comment:
Formatting. Let's not put code within the codeblock on the same line.

This `if` will invalidate the correct result of the previous scanf for `+CREG: <n>,<stat>` responses, causing `stat` to be parsed as `<n>` instead in the second `scanf`.

Reply:
same as above